### PR TITLE
planner: don't decorrelate the APPLY when the inner's projection reference no column

### LINKED ISSUE #32089: Wrong results with correlated subquery
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
drop table if exists t1,t2;
CREATE TABLE t1(a INT PRIMARY KEY);
CREATE TABLE t2(b INT,c INT);
INSERT INTO t1 VALUES (1), (2);
INSERT INTO t2 VALUES (1,2), (2,3);
SELECT (SELECT 1 FROM t1 WHERE a=b AND c=1) FROM t2;
```
### 2. What did you expect to see? (Required)
```sql
mysql> SELECT (SELECT 1 FROM t1 WHERE a=b AND c=1) FROM t2;
+--------------------------------------+
| (SELECT 1 FROM t1 WHERE a=b AND c=1) |
+--------------------------------------+
|                                 NULL |
|                                 NULL |
+--------------------------------------+
2 rows in set (0.00 sec)
```
### 3. What did you see instead (Required)
```sql
mysql> SELECT (SELECT 1 FROM t1 WHERE a=b AND c=1) FROM t2;
+--------------------------------------+
| (SELECT 1 FROM t1 WHERE a=b AND c=1) |
+--------------------------------------+
|                                    1 |
|                                    1 |
+--------------------------------------+
2 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```sql
| Release Version: v5.5.0-alpha-168-g23f7e51ae
Edition: Community
Git Commit Hash: 23f7e51ae01287fa7f811e1462d8987e4d7727a4
Git Branch: master
UTC Build Time: 2022-01-27 17:39:21
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by winoros**
…rences no column

### What problem does this PR solve?

Issue Number: close #32089 

Problem Summary:

When APPLY is left outer join and the inner child is projection without any reference to its child. We cannot decorrelate such APPLY for now. Since the evaluation of the constant expression would be different before and after the decorrelation.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/rule_decorrelate.go [hunk 1]
```diff
@@ -154,6 +154,21 @@ func (s *decorrelateSolver) optimize(ctx context.Context, p LogicalPlan, opt *lo
...
+				if len(expression.ExtractCorColumns(expr)) > 0 || !expression.ExtractColumnSet(expr).IsEmpty() {
+					allConst = false
+					break
+				}
+			}
+			if allConst && apply.JoinType == LeftOuterJoin {
+				// If the projection just references some constant. We cannot directly pull it up when the APPLY is an outer join.
+				//  e.g. select (select 1 from t1 where t1.a=t2.a) from t2; When the t1.a=t2.a is false the join's output is NULL.
+				//       But if we pull the projection upon the APPLY. It will return 1 since the projection is evaluated after the join.
+				// We disable the docorrelation directly for now.
```
Comment (time-and-fate) [inline planner/core/rule_decorrelate.go:168]:
```suggestion
				// We disable the decorrelation directly for now.
```

**Review by qw4990 — APPROVED** [review]

**Review by AilinKid — COMMENTED** [review]
/LGTM

**Pull request merged**